servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultCompositeCloseable.java

```diff
@@ -112,39 +125,85 @@ public void closeGracefully() {
+        final Operand operand;
+        if (operands.isEmpty()) {
+            operand = new Operand(isMerge);
+            operands.add(operand);
+        } else {
+            Operand rawOperand = append ? operands.getLast() : operands.getFirst();
+            if (isMerge == rawOperand.isMerge) {
+                operand = rawOperand;
+            } else {
+                operand = new Operand(isMerge);
+                if (append) {
+                    operands.add(operand);
+                } else {
+                    operands.addFirst(operand);
+                }
+            }
+        }
+        return operand;
     }
 
     private void concatCloseableDelayError(final AsyncCloseable closeable) {
-        closeAsync = closeAsync.concat(closeable.closeAsync().onErrorComplete(th -> {
-            //TODO: This should use concatDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }));
-        closeAsyncGracefully = closeAsyncGracefully.concat(closeable.closeAsyncGracefully().onErrorComplete(th -> {
-            //TODO: This should use concatDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }));
+        final Operand operand = getOrAddConcatOperand(true);
+        operand.closables.add(closeable);
+        resetState();
     }
 
     private void prependCloseableDelayError(final AsyncCloseable closeable) {
-        closeAsync = closeable.closeAsync().onErrorComplete(th -> {
-            //TODO: This should use prependDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }).concat(closeAsync);
-        closeAsyncGracefully = closeable.closeAsyncGracefully().onErrorComplete(th -> {
-            //TODO: This should use prependDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }).concat(closeAsyncGracefully);
+        final Operand operand = getOrAddConcatOperand(false);
+        operand.closables.add(closeable);
+        resetState();
+    }
+
+    private Completable buildCompletable(Function<AsyncCloseable, Completable> closerFunc) {
+        Completable result = completed();
+        for (Operand operand : operands) {
+            if (operand.isMerge) {
+                result = result.mergeDelayError(operand.closables.stream().map(closerFunc).toArray(Completable[]::new));
+            } else {
+                result = result.concat(operand.closables.stream().map(closerFunc).map(completable ->
+                        completable.onErrorComplete(th -> {
+                            // TODO: This should use concatDelayError when available.
+                            LOGGER.debug("Ignored failure to close", th);
+                            return true;
```

Comment:
That was pre-existing, but looks like we hide all errors instead of delaying them. Should we try to save them in `AsyncContext` and verify its key at the end, as the last completable?

Reply:
Lets defer for `concatDelayError` in a followup PR when it is available.